### PR TITLE
chore/implement categorical sampling in rune

### LINKED ISSUE #79: Implement `Rune.Rng.categorical`
## Why it matters
Outreachy applicants will want to sample from categorical distributions when building small language or vision demos. Right now `Rune.Rng.categorical` just raises `failwith`, so any newcomer who calls it hits a runtime crash.

## How to see the gap
Open `rune/lib/rng.ml` and look at the big TODO block above the function definition (lines ~84-105). The comment already sketches the algorithm: softmax logits, take a cumulative sum, compare against uniform samples, pick the first bucket. There is also a permanent `failwith` in `rune/lib/rng.mli`'s documentation that highlights the missing implementation.

## Your task
- Replace the `failwith` in `rune/lib/rng.ml` with a real implementation that follows the pseudo-code comment.
- Make sure the axis argument works for both positive and negative values (e.g. `axis = -1`).
- Add Alcotest coverage beside the other RNG tests in `rune/test/test_rng.ml` so the behavior is verified.

## Tips
- Reuse helpers that already exist on `Rune.Tensor`, such as `softmax`, `cumsum`, `cmplt`, and `argmax`.
- You can borrow the key splitting pattern from the other sampling functions to keep determinism.
- When writing the test, create a tiny logits tensor where you can compute the expected probabilities by hand.

## Done when
- `Rune.Rng.categorical` returns indices with the expected distribution for a fixed seed.
- The new test passes under `dune exec rune/test/test_rng.exe`.
- No other RNG tests regress.

### PR DESCRIPTION
**Pull request by nirnayroy**
fixes #79

### CODE REVIEW

**Comment by tmattio** [comment]
Thanks for the PR @nirnayroy! It seems your committing from another account, that's ok if this is normal, but just wanted to flag that, if that's unintentional, you might want to update your git config to use the right GitHub account.

I'll review the PR soon.

**Comment by tmattio** [comment]
Thanks for taking this on, it's close, but there’s a few issues to fix before merge.

**Fix the sampler to use the Gumbel-max trick**

You're drawing one uniform per class, compare each against the CDF, and taking the `argmax`.
Because those uniforms are independent, the distribution becomes skewed (probabilities like `(1 − p0) * (p0 + p1)` instead of the softmax value `p1`).
So the distribution of our samples is biased and doesn't match the softmax of the logits.

The textbook categorical sampler is "draw exactly one uniform in [0,1), slot it into the CDF, and pick
the first bucket whose cumulative probability exceeds it." We still want that behaviour, but in our code we need a vectorized/JIT-friendly formulation, so the approach you took of drawing one uniform per class and comparing each against the CDF is a good start.

To fix the bias, we can use the Gumbel-max trick: Instead of comparing the uniform draws directly against the CDF, we transform them with `g = -log(-log(U))` (where `U` is the uniform sample) and add them to the logits. This transformation is  just an element-wise operation, so it preserves the order of the samples, but it has the property that when you take `argmax(logits + gumbel)`, you get a sample from the softmax distribution.

**Match JAX semantics**

JAX’s random.categorical keeps logits in their incoming dtype (no forced float32 downcast) and accepts a shape argument so you can draw multiple iid samples per batch without splitting the key manually. Right now we throw away precision by always casting to Float32, and callers can only get one draw per batch element. We should remove the dtype cast and add the optional shape prefix so the API matches JAX more closely.

**Tighten the distribution test**

The test is a good start, but it should assert that every bucket is close to its theoretical probability, not just "the top class is big." After you switch to the Gumbel implementation, keep the full-histogram check so we’ll immediately catch any future regressions in the sampler’s distribution.

Here's some pseudocode with these fixes:

```
  categorical(key, logits, axis=-1, shape=()):
      # 1. Canonicalize the axis and peel apart batch vs class dims.
      axis = canonicalize_axis(axis, logits.ndim)
      batch_shape = logits.shape[:axis] + logits.shape[axis+1:]
      n_classes = logits.shape[axis]

      # 2. Allow callers to request multiple iid samples (same as JAX’s shape=()).
      #    The output will have shape = shape + batch_shape.
      if shape:
          logits = broadcast_to(shape + logits.shape, logits)
          axis = axis + len(shape)  # axis shifts once we prefix extra dims

      # 3. Draw a *single* uniform sample per logits entry, convert to Gumbel noise.
      u = uniform(key, dtype=logits.dtype, shape=logits.shape)
      gumbel = -log(-log(u))

      # 4. Add the noise to the logits and argmax along the class axis.
      noisy_logits = logits + gumbel
      indices = argmax(noisy_logits, axis=axis, keepdims=False)

      # 5. Result already has the desired shape (shape + batch_shape).
      return indices.astype(Int32)
```

**Comment by nirnayroy** [comment]
Hi @tmattio, thanks for the review , I have addressed the config file changes, added support for all types and tightened the distribution test. Please have a look again

**Comment by tmattio** [comment]
Thanks @nirnayroy, I've added another test and reworked the dtype handling slightly to have higher precision on float64, it looked very good otherwise, merged!
